### PR TITLE
[KYUUBI #886] Add HTTP transport mode support to KYUUBI

### PR DESCRIPTION
**Pull request by mahmoudbahaa**
### _Why are the changes needed?_

Add http transport protocol support to Kyuubi
main code was taken from:

1. https://github.com/SteNicholas/incubator-kyuubi/commit/25ac0731d7331616988026996ca918793af151e9
2. https://github.com/apache/hive/blob/branch-3.1/service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpCLIService.java
2. https://github.com/apache/hive/blob/branch-3.1/service/src/java/org/apache/hive/service/cli/thrift/ThriftHttpServlet.java

### _How was this patch tested?_
- [ ] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate
Tested against NONE and LDAP,
Not tested against KERBEROS or CUSTOM
Screenshot_20220604_133700
Screenshot_20220604_133739
Screenshot_20220604_133811
Screenshot_20220604_133934
Screenshot_20220604_134032

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by ulysses-you** [comment]
thank you @mahmoudbahaa working on this ! cc @turboFei @yanghua @yaooqinn

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+      val schemeName = if (useSsl) "https" else "http"
+
+      // Change connector if SSL is used
+      val connector =
+        if (useSsl) {
+          val keyStorePath = conf.get(FRONTEND_THRIFT_HTTP_SSL_KEYSTORE_PATH)
+
+          if (keyStorePath.isEmpty) {
+            throw new IllegalArgumentException(FRONTEND_THRIFT_HTTP_SSL_KEYSTORE_PATH.key +
+              " Not configured for SSL connection")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:110]: Can we point out the missed information detail of the config item?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+          }
+
+          val keyStorePassword = conf.get(FRONTEND_THRIFT_HTTP_SSL_KEYSTORE_PASSWORD)
+          if (keyStorePassword.isEmpty) {
+            throw new IllegalArgumentException(FRONTEND_THRIFT_HTTP_SSL_KEYSTORE_PASSWORD.key +
+              " Not configured for SSL connection")
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:116]: ditto

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+          }
+
+          val sslContextFactory = new SslContextFactory.Server
+          val excludedProtocols = conf.get(FRONTEND_THRIFT_HTTP_SSL_PROTOCOL_BLACKLIST).split(",")
+          info("HTTP Server SSL: adding excluded protocols: " +
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:121]: More readable log information? Kyuubi has owned an HTTP server to serve restful API. So let's make the `HTTP server` clear?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 4]
```diff
@@ -0,0 +1,285 @@
...
+      val context = new ServletContextHandler(ServletContextHandler.SESSIONS)
+      context.setContextPath("/")
+
+      context.addEventListener(new ServletContextListener() {
+        override def contextInitialized(servletContextEvent: ServletContextEvent): Unit = {
+          val metrics = MetricsFactory.getInstance
+          if (metrics != null) {
+            try {
+              metrics.incrementCounter(MetricsConstant.OPEN_CONNECTIONS)
+              metrics.incrementCounter(MetricsConstant.CUMULATIVE_CONNECTION_COUNT)
```
Comment (yanghua) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:159]: Glad to see we report some metrics here. Can we also make sure if it is suitable for Kyuubi's metric system?

**Review by yanghua — COMMENTED** [review]
@mahmoudbahaa Thanks for your contribution! Left some comments.

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 5]
```diff
@@ -434,6 +452,121 @@ object KyuubiConf {
...
 
+  val FRONTEND_THRIFT_HTTP_REQUEST_HEADER_SIZE: ConfigEntry[Int] =
+    buildConf("kyuubi.frontend.thrift.http.request.header.size")
+      .doc("Request header size in bytes, when using HTTP transport mode. Jetty defaults used.")
+      .version("1.6.0")
+      .intConf
+      .createWithDefault(6 * 1024)
+
+  val FRONTEND_THRIFT_HTTP_RESPONSE_HEADER_SIZE: ConfigEntry[Int] =
+    buildConf("frontend.thrift.http.response.header.size")
```
Comment (pan3793) [inline kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala:463]: "frontend.xxx" => "kyuubi.frontend.xxx"

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+          new ServerConnector(
+            server.get,
+            sslContextFactory,
+            connectionFactory)
+        } else {
+          new ServerConnector(server.get, connectionFactory)
+        }
+
+      connector.setPort(portNum)
+      // Linux:yes, Windows:no
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:140]: maybe Linux => Unix-like ?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+      } else {
+        server.foreach(_.setHandler(context))
+      }
+
+      context.addServlet(new ServletHolder(servlet), httpPath)
+      constrainHttpMethods(context)
+
+      info("Started " + getClass.getSimpleName +
+        " in " + schemeName + " mode on port " + portNum +
+        " path=" + httpPath + " with " + minThreads + "..." + maxThreads + " worker threads")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:188]: let's use s"text $var1 ${var2.var3}"

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+    } catch {
+      case e: Throwable =>
+        MetricsSystem.tracing(_.incCount(THRIFT_CONN_FAIL))
+        error(e)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:192]: redundant?

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 9]
```diff
@@ -0,0 +1,278 @@
...
+          s"Failed to initialize frontend service on $serverAddr:$portNum.",
+          e)
+    }
+    super.initialize(conf)
+  }
+
+  override def run(): Unit =
+    try {
+      if (isServer()) {
+        info(s"Starting and exposing JDBC connection at: jdbc:hive2://$connectionUrl/")
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:203]: Beeline default use thrift binary mode, we need another parameters for thrift http mode

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 10]
```diff
@@ -0,0 +1,278 @@
...
+
+  private def getShortName(userName: String): String = {
+    var ret: String = null
+
+    if (userName != null) {
+      if (authFactory.isKerberosEnabled) {
+        // KerberosName.getShorName can only be used for kerberos user, but not for the user
+        // logged in via other authentications such as LDAP
+        val fullKerberosName = ShimLoader.getHadoopShims.getKerberosNameShim(userName)
+        ret = fullKerberosName.getShortName
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:269]: question: kyuubi support enable kerberos and LDAP in same time, is it sufficient to handle that case?

File: .gitignore [hunk 11]
```diff
@@ -74,3 +74,6 @@ embedded_zookeeper/
 **/operation_logs/
 **/server_operation_logs/
 **/engine_operation_logs/
+conf/log4j2.properties
+conf/kyuubi-defaults.conf
+conf/kyuubi-env.sh
```
Comment (pan3793) [inline .gitignore:79]: I think it's a good idea to exclude those files then we can debug with specific configurations in local easily

File: dev/reformat [hunk 12]
```diff
@@ -22,4 +22,4 @@ KYUUBI_HOME="$(cd "`dirname "$0"`/.."; pwd)"
 
 PROFILES="-Pflink-provided,hive-provided,spark-provided,spark-block-cleaner,spark-3.3,spark-3.2,spark-3.1,spark-3.0,tpcds"
 
-${KYUUBI_HOME}/build/mvn spotless:apply $PROFILES
+mvn spotless:apply $PROFILES
```
Comment (pan3793) [inline dev/reformat:25]: Please revert this change when this PR get ready, the `${KYUUBI_HOME}/build/mvn` will pick up the local `mvn` if there is apposite `mvn` in your PATH rather than download a new one

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+      // Context handler
+      val context = new ServletContextHandler(ServletContextHandler.SESSIONS)
+      context.setContextPath("/")
+
+      context.addEventListener(new ServletContextListener() {
+        override def contextInitialized(servletContextEvent: ServletContextEvent): Unit = {
+          MetricsSystem.tracing { ms =>
+            ms.incCount(THRIFT_CONN_TOTAL)
+            ms.incCount(THRIFT_CONN_OPEN)
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:160]: we should inc/dec it in thrift binary as well or just change the name

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/ThriftHttpServlet.scala [hunk 14]
```diff
@@ -0,0 +1,539 @@
...
+  val AUTH_COOKIE = "hive.server2.auth"
+  private val RAN = new SecureRandom()
+  private var isCookieAuthEnabled: Boolean = false
+  private var cookieDomain: String = _
+  private var cookiePath: String = _
+  private var cookieMaxAge: Int = 0
+  private var isCookieSecure = false
+  private var isHttpOnlyCookie = false
+  private val HIVE_DELEGATION_TOKEN_HEADER = "X-Hive-Delegation-Token"
+  private val X_FORWARDED_FOR = "X-Forwarded-For"
```
Comment (pan3793) [inline kyuubi-server/src/main/scala/org/apache/kyuubi/server/http/ThriftHttpServlet.scala:70]: Please unify the name style of `HIVE_DELEGATION_TOKEN_HEADER` and `X_FORWARDED_FOR`

File: kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala (hunk 9 above)
Comment (mahmoudbahaa) [reply kyuubi-server/src/main/scala/org/apache/kyuubi/server/KyuubiTHttpFrontendService.scala:203]: hhtp mode is enabled in beeline/jdbc using the `transportMode=http` option for example `jdbc:hive2://localhost:10010/default;transportMode=http`

**Comment by mahmoudbahaa** [comment]
I made a new PR with no kerberos support for now (can be added at later date) as it is not yet feature complete nor tested so I will be closing this PR and we can continue discussion on the other PR https://github.com/apache/incubator-kyuubi/pull/2833
